File: context_tree_cli/rules/ci_validation.py
"""Rule: check CI/CD validation workflows."""

from __future__ import annotations

from context_tree_cli.repo import Repo


def evaluate(repo: Repo) -> dict:
    tasks = []
    has_validation = False
    workflows_dir = repo.root / ".github" / "workflows"
    if workflows_dir.is_dir():
        for wf in workflows_dir.iterdir():
            if wf.is_file() and wf.suffix in (".yml", ".yaml"):
                try:
                    content = wf.read_text()
                    if "validate_nodes" in content or "validate_members" in content:
                        has_validation = True
                        break
                except OSError:
                    continue
    if not has_validation:
        tasks.append(
            "No validation workflow found — copy"
            " `.context-tree/workflows/validate.yml` to"
            " `.github/workflows/validate.yml`"
        )
    return {"group": "CI / Validation", "order": 6, "tasks": tasks}

File: src/rules/ci-validation.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import type { Repo } from "../repo.js";
import type { RuleResult } from "./index.js";

export function evaluate(repo: Repo): RuleResult {
  const tasks: string[] = [];
  let hasValidation = false;
  const workflowsDir = join(repo.root, ".github", "workflows");
  try {
    if (statSync(workflowsDir).isDirectory()) {
      for (const name of readdirSync(workflowsDir)) {
        if (!name.endsWith(".yml") && !name.endsWith(".yaml")) continue;
        const fullPath = join(workflowsDir, name);
        try {
          if (!statSync(fullPath).isFile()) continue;
          const content = readFileSync(fullPath, "utf-8");
          if (
            content.includes("validate_nodes") ||
            content.includes("validate_members")
          ) {
            hasValidation = true;
            break;
          }
        } catch {
          continue;
        }
      }
    }
  } catch {
    // workflows dir doesn't exist
  }
  if (!hasValidation) {
    tasks.push(
      "No validation workflow found — copy `.context-tree/workflows/validate.yml` to `.github/workflows/validate.yml`",
    );
  }
  return { group: "CI / Validation", order: 6, tasks };
}
